components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java

```diff
@@ -1505,6 +1510,93 @@ public static SpOAuth2ExpiryTimeConfiguration getSpTokenExpiryTimeConfig(String
         return spTokenTimeObject;
     }
 
+    /**
+     * Retrieve audience configured for the particular service provider.
+     *
+     * @param clientId
+     * @param tenantDomain
+     * @return
+     */
+    public static List<String> getOIDCAudience(String clientId, String tenantDomain) {
+
+        List<String> oidcAudiences = getDefinedCustomOIDCAudiences(clientId, tenantDomain);
+        // Need to add client_id as an audience value according to the spec.
+        if (!oidcAudiences.contains(clientId)) {
+            oidcAudiences.add(0, clientId);
+        } else {
+            Collections.swap(oidcAudiences, oidcAudiences.indexOf(clientId), 0);
+        }
+        return oidcAudiences;
+    }
+
+    private static List<String> getDefinedCustomOIDCAudiences(String clientId, String tenantDomain) {
+
+        List<String> audiences = new ArrayList<>();
+
+        // Priority should be given to service provider specific audiences over globally configured ones.
+        if (OAuth2ServiceComponentHolder.isAudienceEnabled()) {
+            OIDCAudienceCacheKey cacheKey = new OIDCAudienceCacheKey(clientId + "@" + tenantDomain);
```

Comment:
client_id is unique across tenant domains. So should we use tenantDomain in the key?

Reply:
@mefarazath this seems to be a refactored method which was originally in DefaultIDTokenBuilder.java. So let's keep that part as it is.